keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,604 @@
+* Volume QoS classes. This proposal is for fine-grained provisioned IO control, and not for defining general classes of volume performance. Currently, storage vendors may accomplish this by adding custom parameters to a storage class. A future KEP may codify such behavior with common storage class parameter names, but that is out of scope of this KEP.
+* Scheduling based on total IO limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+Add two well-known ResourceName keys to the existing ResourceList type used in PersistentVolumeClaim.Spec.Resources.{Limits,Requests} and PersistentVolume.Spec.Capacity:
+* (ReadWrite) IOPS independent from capacity
+* (ReadWrite) Throughput
+These values can be changed on a PersistentVolumeClaim, and will be reconciled against the storage provider in the same way as storage capacity resizing. The status of a PVC will reflect the currently provisioned IO as reported by the storage provider (in the same way that the status states the current size of the volume during a resize operation).
+
+```
+// Example:
+kind: PersistentVolumeClaim
+...
+spec:
+  ...
+  resources:
+requests:
+  storage: 4Gi
+  throughput: "1000M" // MiB/s as unit
+  iops: "160000"     // IO/s as unit
+// Vendors may add additional resources using Device Plugin syntax: vendor-domain/resourcetype
+```
+
+[Device Plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) is out of scope for this KEP and it is one of the options for configure vendor specific storage options.
+
+Add PVC Status field; change ResizeStatus alpha field to AllocatedResourceStatus map.
+
+### CSI API
+The CSI create request will be extended to add provisioned IO parameters. For volume creation, cloud providers can add iops and throughput field in parameters and process in the csi driver, an example:
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gcepd
+provisioner: pd.csi.storage.gke.io
+parameters:
+  type: hyperdisk-balanced
+  provisioned-iops-on-create: '10000'
+  provisioned-throughput-on-create: '1000Mi'
+volumeBindingMode: WaitForFirstConsumer
+```
+
+A new ControllerModifyVolume RPC will be added. The current spec requires ["If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK"](https://github.com/container-storage-interface/spec/blob/master/spec.md). The change here will be if the capacity change is for iops and throughput, we will not require that to be larger than the current configuration.
+
+#### ControllerModifyVolume
```

Comment:
Would the driver need to return new capabilities to specify if it can support this?
I imagine we would need a capability for controller modify volume but also a way to know if the driver supports iops and/or throughput. Maybe the latter is the exceeds capabilities error?